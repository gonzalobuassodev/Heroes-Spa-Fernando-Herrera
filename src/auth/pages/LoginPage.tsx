import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';

export const LoginPage = () => {

    const { login } = useContext(AuthContext);
    const navigate = useNavigate();


    const onLogin = () => {
        const lastPath = localStorage.getItem('lastPath') || '/';
        // console.log(lastPath);

        login('2','Gonzalo Buasso');
        navigate(lastPath, { replace: true });
    };
    return (
        <div className="px-4">
            <h1 className="py-4 text-3xl">Login</h1>
            <hr />

            <button
                onClick={onLogin}
                className="bg-blue-500 hover:bg-blue-800 text-white p-2 rounded-md m-4"
            >
                Login
            </button>
        </div>
    );
};
