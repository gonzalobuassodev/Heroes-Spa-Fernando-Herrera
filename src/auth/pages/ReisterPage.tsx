
export const ReisterPage = () => {
  return (
    <div>ReisterPage</div>
  )
}
