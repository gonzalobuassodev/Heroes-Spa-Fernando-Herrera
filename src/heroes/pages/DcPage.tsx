import { HeroesList } from "../components";
import { HeroEnum } from "../Enum/hero.enum";

export const DcPage = () => {

  return (
      <>
          <h1>DcPage</h1>
          <hr />

          <HeroesList publisher={HeroEnum.DC} />
          
      </>
  );
}
