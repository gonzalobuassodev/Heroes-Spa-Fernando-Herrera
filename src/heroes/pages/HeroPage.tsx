import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroesById } from '../helpers';
import { useMemo } from 'react';

export const HeroPage = () => {
    const { id } = useParams();

    const navigate = useNavigate();

    // console.log(id)
    const hero = useMemo(() => getHeroesById(id), [id]); ;

    const onNavigateBack = () => {
      navigate(-1);
    }

    if (!hero) {
        return <Navigate to="/marvel" />;
    }


    // console.log(hero)
    return (
        <div className="px-4 mt-5">
            <div className="grid grid-cols-2 gap-4">
                <img
                    className="animate__animated animate__fadeInLeft"
                    src={`/assets/heroes/${hero.id}.jpg`}
                    alt={hero.id}
                />
                <div>
                    <h1 className="text-3xl">{hero.superhero}</h1>
                    <ul>
                        <li className="ml-5 my-2">
                            <b>Alter ego:</b> {hero.alter_ego}
                        </li>
                        <li className="ml-5 my-2">
                            <b>Publisher:</b> {hero.publisher}
                        </li>
                        <li className="ml-5 my-2">
                            <b>First appearence:</b> {hero.first_appearance}
                        </li>
                    </ul>

                    <h5 className="mt-3 text-xl">Characters</h5>
                    <p>{hero.characters}</p>

                    <button
                        onClick={onNavigateBack}
                        className="border-blue-500 border rounded-md p-3 mt-3 hover:bg-blue-800 hover:text-white"
                    >
                        Regresar
                    </button>
                </div>
            </div>
        </div>
    );
};
