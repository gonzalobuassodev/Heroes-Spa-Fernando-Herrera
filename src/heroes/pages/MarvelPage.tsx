import { HeroesList } from '../components';
import { HeroEnum } from '../Enum/hero.enum';

export const MarvelPage = () => {
    return (
        <div className='px-4 mt-5'>
            <h1>Marvel</h1>
            <hr />

            <HeroesList publisher={HeroEnum.MARVEL} />
        </div>
    );
};
