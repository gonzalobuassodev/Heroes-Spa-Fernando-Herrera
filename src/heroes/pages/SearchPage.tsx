import { FormEvent } from 'react';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { useLocation, useNavigate } from 'react-router-dom';
import { getHeroesByName } from '../helpers';
import { IHero } from '../interfaces';
import { HeroCard } from '../components';

interface QueryParams {
    q?: string;
}

export const SearchPage = () => {
    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search) as QueryParams;

    const heroes = getHeroesByName(q);

    const showSearch = q === '';
    const showErrors = heroes.length < 1 && q !== '';

    const { formState, onInputChange } = useForm({
        searchText: q,
    });

    const { searchText } = formState;

    const onSearchSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        navigate(`?q=${searchText.toLocaleLowerCase().trim()}`);
    };

    return (
        <>
            <h1>Search</h1>
            <hr />
            <div className="grid grid-cols-2 gap-2">
                <div>
                    <h4>Searching</h4>

                    <hr />

                    <form aria-label="form" onSubmit={(e) => onSearchSubmit(e)}>
                        <input
                            type="text"
                            placeholder="Search a Hero"
                            className="border-2 border-slate-500 rounded-md p-2"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={onInputChange}
                        />

                        <button className="bg-blue-500 hover:bg-blue-800 text-white p-2 rounded-md m-4">
                            Search
                        </button>
                    </form>
                </div>

                <div>
                    <h4>Results</h4>
                    <hr />

                    <div
                        aria-label="alert-search"
                        className={`mt-3 p-4 bg-blue-300 rounded-md border-2 border-blue-500 
                          ${
                              showSearch ? 'block' : 'hidden'
                          } animate__animated animate__fadeIn`}
                    >
                        Search a Hero
                    </div>

                    <div
                        aria-label="alert-no-results"
                        className={`my-3 p-4 bg-pink-200 rounded-md border-2 border-slate-500
                          ${
                              showErrors ? 'block' : 'hidden'
                          } animate__animated animate__fadeIn`}
                    >
                        There`s no results with <b>{q}</b>
                    </div>

                    {heroes &&
                        heroes.map((hero: IHero) => (
                            <HeroCard key={hero.id} hero={hero} />
                        ))}
                </div>
            </div>
        </>
    );
};
