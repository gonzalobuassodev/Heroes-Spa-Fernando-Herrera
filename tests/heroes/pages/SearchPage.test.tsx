import React from 'react';
import { render } from '@testing-library/react';
import { fireEvent, screen } from '@testing-library/dom';
import { MemoryRouter } from 'react-router-dom';

import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}));

describe('Testing SearchPage', () => {
    beforeEach(() => jest.clearAllMocks);
    test('should show the components with the default values', () => {
        const { container } = render(
            <MemoryRouter
                future={{
                    v7_relativeSplatPath: true,
                    v7_startTransition: true,
                }}
            >
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
        // screen.debug();
    });

    test('should show to Batman and the input value with the query params value', () => {
        render(
            <MemoryRouter
                initialEntries={['/search?q=batman']}
                future={{
                    v7_relativeSplatPath: true,
                    v7_startTransition: true,
                }}
            >
                <SearchPage />
            </MemoryRouter>
        );

        const inputSearch = screen.getByRole('textbox') as HTMLInputElement;
        expect(inputSearch.value).toBe('batman');

        const img = screen.getByRole('img') as HTMLImageElement;
        expect(img.src).toContain('batman');

        const alertSearch = screen.getByLabelText('alert-search');
        expect(alertSearch.className).toContain('hidden');

        const alertNoResult = screen.getByLabelText('alert-no-results');
        expect(alertNoResult.className).toContain('hidden');

        // screen.debug();
    });

    test('should show an error if dont find the hero', () => {
        render(
            <MemoryRouter
                initialEntries={['/search?q=batman21334']}
                future={{
                    v7_relativeSplatPath: true,
                    v7_startTransition: true,
                }}
            >
                <SearchPage />
            </MemoryRouter>
        );

        const alertNoResult = screen.getByLabelText('alert-no-results');
        expect(alertNoResult.className).not.toContain('hidden');

        // screen.debug();
    });

    test('should call navigate function to a new page', () => {

        const inputValue = 'superman';
        
        render(
            <MemoryRouter
                initialEntries={['/search']}
                future={{
                    v7_relativeSplatPath: true,
                    v7_startTransition: true,
                }}
            >
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox') as HTMLInputElement;
        fireEvent.change(input, {
            target: { name: 'searchText', value: inputValue },
        });

        // console.log(input.value)

        const form = screen.getByLabelText('form');
        fireEvent.submit(form);

        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
    });
});
