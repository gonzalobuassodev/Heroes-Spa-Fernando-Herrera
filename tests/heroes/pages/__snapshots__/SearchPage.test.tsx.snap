// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchPage should show the components with the default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="grid grid-cols-2 gap-2"
  >
    <div>
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="border-2 border-slate-500 rounded-md p-2"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          class="bg-blue-500 hover:bg-blue-800 text-white p-2 rounded-md m-4"
        >
          Search
        </button>
      </form>
    </div>
    <div>
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alert-search"
        class="mt-3 p-4 bg-blue-300 rounded-md border-2 border-blue-500 
                          block animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="alert-no-results"
        class="my-3 p-4 bg-pink-200 rounded-md border-2 border-slate-500
                          hidden animate__animated animate__fadeIn"
      >
        There\`s no results with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
